src/client.rs

```diff
@@ -1912,31 +1954,107 @@ impl TryFrom<&ChallengeRef<'_>> for BearerChallenge {
-                // Fallback to hashing payload (tested with ECR)
-                let digest = sha2::Sha256::digest(body);
-                let hex = format!("sha256:{:x}", digest);
-                debug!(%hex, "Computed digest of manifest payload.");
-                Ok(hex)
-            } else {
-                Err(OciDistributionError::RegistryNoDigestError)
+/// Helper for extracting `Docker-Content-Digest` header from manifest GET or HEAD request.
+fn digest_header_value(headers: HeaderMap) -> Result<Option<String>> {
+    headers
+        .get(DOCKER_DIGEST_HEADER)
+        .map(|hv| hv.to_str().map(|s| s.to_string()))
+        .transpose()
+        .map_err(|e| {
+            OciDistributionError::GenericError(Some(format!(
+                "Cannot convert Docker-Content-Digest header value to string: {}",
+                e
+            )))
+        })
+}
+
+/// Given the optional digest header value and digest of the reference, returns the digest of the
+/// content, validating that the digest of the content matches the proper digest
+fn validate_digest(
+    body: &[u8],
+    digest_header: Option<String>,
+    reference_digest: Option<&str>,
+) -> Result<String> {
+    match compare_and_validate_digests(digest_header, reference_digest)? {
+        Some(digest) => calculate_and_validate(body, &digest),
+        // If we have neither, just digest the body
+        None => Ok(sha256_digest(body)),
+    }
+}
+
+/// This function compares the possible digest header value with a digest of a reference and returns
+/// the proper digest if it is valid. If both digests are provided, but they use different
+/// algorithms, then the header digest is returned as according to the spec it should be used for
+/// validation
+fn compare_and_validate_digests(
+    digest_header: Option<String>,
+    reference_digest: Option<&str>,
+) -> Result<Option<String>> {
+    match (digest_header, reference_digest) {
+        (None, None) => Ok(None),
+        (Some(digest), None) => Ok(Some(digest)),
+        (None, Some(digest)) => Ok(Some(digest.to_string())),
+        (Some(digest), Some(ref_digest)) => {
+            // According to the spec, it is just fine if the requested digest differs from the
+            // canonical digest from the headers if the algorithms are different. If they are the
+            // same, then the digests should match.
+            let (algo, _) = digest.split_once(':').ok_or_else(|| {
+                OciDistributionError::GenericError(Some(
+                    "Incorrect digest format on header value. Should have algorithm".to_string(),
+                ))
+            })?;
+            let (ref_algo, _) = ref_digest.split_once(':').ok_or_else(|| {
+                OciDistributionError::GenericError(Some(
+                    "Incorrect digest format on image reference. Should have algorithm".to_string(),
+                ))
+            })?;
+            if algo != ref_algo {
+                return Ok(Some(digest));
```

Comment:
I fear that this still leaves us open to a malicious response that does not match the ref digest.

Browsing the spec a bit more, I think it should be safe to verify both the requested and the canonical digest, because

> A GET request to an existing manifest URL MUST provide the expected manifest, with a response code that MUST be 200 OK. 
> [...]
> If the blob is not found in the registry, the response code MUST be 404 Not Found.

I read that as "the registry is NOT free to return content that does not match the request". So if the requested reference was a digest and the response code is 200, the registry found a manifest with that digest and thus the body should match it.

How about returning a `Vec` here and verifying all (0, 1 or 2) returned digests?

Reply:
It might be a little weird, but I can do it. I'll get back to it on Monday